File: 17-letter-combinations-of-a-phone-number/17-letter-combinations-of-a-phone-number.js
var letterCombinations = function(digits) {
    if(!digits.length) return [];
    
    const map = {
        2: ['a', 'b', 'c'],
        3: ['d', 'e', 'f'],
        4: ['g', 'h', 'i'],
        5: ['j', 'k', 'l'],
        6: ['m', 'n', 'o'],
        7: ['p', 'q', 'r', 's'],
        8: ['t', 'u', 'v'],
        9: ['w', 'x', 'y', 'z'],
    }
    
    const result = [];
    
    function permute(str, idx) {
        if(idx === digits.length) {
            result.push(str);
            return;
        }
        
        for(let alpha of map[digits[idx]]) {
            permute(str+alpha, idx+1);
            /*
             permute("a", 1);
             permute("b", 1);
             permute("c", 1);
            */
        }
    }
    permute('', 0);
    return result;
};

File: 17-letter-combinations-of-a-phone-number/17-letter-combinations-of-a-phone-number.py
class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        if len(digits)==0: return []
        dic={2:"abc",3:"def",4:"ghi",5:"jkl",6:"mno",7:"pqrs",8:"tuv",9:"wxyz"}
        arr=[]    
        def dfs(string,idx):
            if len(string)==len(digits):
                arr.append(string)
                return
            for i in dic[int(digits[idx])]:
                dfs(string+i,idx+1)

        dfs("",0)
        return arr